baseband/guppi/payload.py

```diff
@@ -0,0 +1,239 @@
+        Header that provides information about how the payload is encoded.
+        If not given, the following arguments have to be passed in.
+    bps : int, optional
+        Number of bits per sample part (i.e., per channel and per real or
+        imaginary component).  Default: 8.
+    sample_shape : tuple, optional
+        Shape of the samples; e.g., (nchan,).  Default: ().
+    complex_data : bool, optional
+        Whether data are complex.  Default: `False`.
+    time_ordered : bool, optional
+        Whether data are time-ordered.  Default: `True`.
+    """
+    _decoders = {
+        8: decode_8bit}
+    _encoders = {
+        8: encode_8bit}
+    _dtype_word = np.dtype('int8')
+
+    _sample_shape_maker = namedtuple('SampleShape', 'npol, nchan')
+
+    def __init__(self, words, header=None, sample_shape=(), bps=8,
+                 complex_data=False, time_ordered=True):
+        if header is not None:
+            bps = header.bps
+            sample_shape = header.sample_shape
+            complex_data = header.complex_data
+            time_ordered = header.time_ordered
+        super(GUPPIPayload, self).__init__(words, sample_shape=sample_shape,
+                                           bps=bps, complex_data=complex_data)
+        self.time_ordered = time_ordered
+        # If time ordered, _item_to_slices works on per-channel words.
+        if self.time_ordered:
+            self._bpfs //= self.sample_shape.nchan
+
+    @classmethod
+    def fromfile(cls, fh, header=None, memmap=False, payload_nbytes=None,
+                 **kwargs):
+        """Read or map encoded data in file.
+
+        Parameters
+        ----------
+        fh : filehandle
+            Handle to the file which will be read or mapped.
+        header : `~baseband.dada.GUPPIHeader`, optional
+            If given, used to infer ``payload_nbytes``, ``bps``,
+            ``sample_shape``, ``complex_data`` and ``time_ordered``.  If not
+            given, those have to be passed in.
+        memmap : bool, optional
+            If `False` (default), read from file.  Otherwise, map the file in
+            memory (see `~numpy.memmap`).
+        payload_nbytes : int, optional
+            Number of bytes to read (default: as given in ``header``,
+            ``cls._nbytes``, or, for mapping, to the end of the file).
+        **kwargs
+            Additional arguments are passed on to the class initializer. These
+            are only needed if ``header`` is not given.
+        """
+        if payload_nbytes is None:
+            payload_nbytes = (cls._nbytes if header is None
+                              else header.payload_nbytes)
+
+        if not memmap:
+            return super(GUPPIPayload, cls).fromfile(
+                fh, header=header, payload_nbytes=payload_nbytes, **kwargs)
+
+        if hasattr(fh, 'memmap'):
+            words = fh.memmap(dtype=cls._dtype_word,
+                              shape=None if payload_nbytes is None else
```

Comment:
Can we precalculate the shape, since it is needed in both branches of the `if` statement?

Reply:
Done.